Source/WTF/wtf/Vector.h

```diff
@@ -56,269 +56,152 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
 
 enum class NulloptBehavior : bool { Ignore, Abort };
 
-template <bool needsDestruction, typename T>
-struct VectorDestructor;
-
-template<typename T>
-struct VectorDestructor<false, T>
-{
-    static void destruct(T*, T*) { }
-};
-
 WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
 
 template<typename T>
-struct VectorDestructor<true, T>
-{
-    static void destruct(T* begin, T* end) 
-    {
-        for (T* cur = begin; cur != end; ++cur)
-            cur->~T();
-    }
-};
-
-template <bool needsInitialization, bool canInitializeWithMemset, typename T>
-struct VectorInitializer;
-
-template<bool canInitializeWithMemset, typename T>
-struct VectorInitializer<false, canInitializeWithMemset, T>
-{
-    static void initializeIfNonPOD(T*, T*) { }
-
-    static void initialize(T* begin, T* end)
-    {
-        VectorInitializer<true, canInitializeWithMemset, T>::initialize(begin, end);
-    }
-};
-
-template<typename T>
-struct VectorInitializer<true, false, T>
-{
-    static void initializeIfNonPOD(T* begin, T* end)
-    {
-        for (T* cur = begin; cur != end; ++cur)
-            new (NotNull, cur) T();
-    }
-
-    static void initialize(T* begin, T* end)
-    {
-        initializeIfNonPOD(begin, end);
-    }
-
-    template<typename... Args>
-    static void initializeWithArgs(T* begin, T* end, Args&&... args)
-    {
-        for (T *cur = begin; cur != end; ++cur)
-            new (NotNull, cur) T(args...);
-    }
-};
-
-template<typename T>
-struct VectorInitializer<true, true, T>
-{
-    static void initializeIfNonPOD(T* begin, T* end)
-    {
-        memset(static_cast<void*>(begin), 0, reinterpret_cast<char*>(end) - reinterpret_cast<char*>(begin));
-    }
-
-    static void initialize(T* begin, T* end)
-    {
-        initializeIfNonPOD(begin, end);
-    }
-};
-
-template <bool canMoveWithMemcpy, typename T>
-struct VectorMover;
-
-template<typename T>
-struct VectorMover<false, T>
-{
-    static void move(T* src, T* srcEnd, T* dst)
-    {
-        while (src != srcEnd) {
-            new (NotNull, dst) T(WTFMove(*src));
-            src->~T();
-            ++dst;
-            ++src;
-        }
-    }
-    static void moveOverlapping(T* src, T* srcEnd, T* dst)
+struct VectorCopier {
+    template<typename U>
+    static void uninitializedCopy(const U* src, const U* srcEnd, T* dst)
     {
-        if (src > dst)
-            move(src, srcEnd, dst);
+        if constexpr (std::is_trivial_v<T> && std::is_same_v<T, U>)
```

Comment:
Should we consider `std::same_as` instead of `std::is_same_v`? Kind of annoying that C++ has both.